File: БР1.2/Владзиевский Евгений/labs/lab1/internal/services/auth_service.go
package services

import (
	"errors"
	"rental-api/internal/config"
	"rental-api/internal/models"
	"rental-api/internal/repositories"
	"rental-api/internal/utils"
)

type AuthService interface {
	Register(input RegisterInput) (*models.User, string, string, error)
	Login(email, password string) (*models.User, string, string, error)
}

type authService struct {
	userRepo repositories.UserRepository
	cfg      *config.Config
}

type RegisterInput struct {
	Email    string
	Password string
	FullName string
	Phone    *string
	Role     string
}

func NewAuthService(userRepo repositories.UserRepository, cfg *config.Config) AuthService {
	return &authService{userRepo: userRepo, cfg: cfg}
}

func (s *authService) Register(input RegisterInput) (*models.User, string, string, error) {
	// Check if user already exists
	existing, err := s.userRepo.FindByEmail(input.Email)
	if err == nil && existing != nil {
		return nil, "", "", errors.New("user with this email already exists")
	}

	// Hash password
	hashedPassword, err := utils.HashPassword(input.Password)
	if err != nil {
		return nil, "", "", err
	}

	// Create user
	user := &models.User{
		Email:        input.Email,
		PasswordHash: hashedPassword,
		FullName:     input.FullName,
		Phone:        input.Phone,
		Role:         input.Role,
	}
	if user.Role == "" {
		user.Role = "tenant"
	}

	err = s.userRepo.Create(user)
	if err != nil {
		return nil, "", "", err
	}

	// Generate tokens
	accessToken, refreshToken, err := utils.GenerateTokenPair(user, s.cfg.JWTSecret)
	if err != nil {
		return nil, "", "", err
	}

	return user, accessToken, refreshToken, nil
}

func (s *authService) Login(email, password string) (*models.User, string, string, error) {
	// Find user by email
	user, err := s.userRepo.FindByEmail(email)
	if err != nil {
		return nil, "", "", errors.New("invalid email or password")
	}

	// Verify password
	if !utils.CheckPasswordHash(password, user.PasswordHash) {
		return nil, "", "", errors.New("invalid email or password")
	}

	// Generate tokens
	accessToken, refreshToken, err := utils.GenerateTokenPair(user, s.cfg.JWTSecret)
	if err != nil {
		return nil, "", "", err
	}

	return user, accessToken, refreshToken, nil
}

File: БР1.2/Москалец Данила/labs/lab1/sangel-api/src/modules/auth/auth.service.ts
import { Repository } from 'typeorm';
import { User, UserRole } from '../user/user.entity';
import { AppDataSource } from '../../config/database';
import { hashPassword } from '../../utils/hash-password';
import { checkPassword } from '../../utils/check-password';
import { generateAccessToken, generateRefreshToken, JwtPayload } from '../../utils/jwt';
import { RegisterDto, LoginDto, TokenPairResponse, UserResponse } from './auth.dto';

export class AuthService {
  private userRepository: Repository<User>;

  constructor() {
    this.userRepository = AppDataSource.getRepository(User);
  }

  private toUserResponse(user: User): UserResponse {
    return {
      id: user.id,
      email: user.email,
      role: user.role,
      first_name: user.first_name,
      last_name: user.last_name,
      middle_name: user.middle_name,
      is_verified: user.is_verified,
      created_at: user.created_at,
    };
  }

  private generateTokens(user: User): TokenPairResponse {
    const payload: JwtPayload = {
      userId: user.id,
      email: user.email,
      role: user.role,
    };

    return {
      access_token: generateAccessToken(payload),
      refresh_token: generateRefreshToken(payload),
      expires_in: 900, // 15 minutes in seconds
    };
  }

  async register(dto: RegisterDto): Promise<UserResponse> {
    // Check if user already exists
    const existingUser = await this.userRepository.findOne({
      where: { email: dto.email },
    });

    if (existingUser) {
      throw new Error('User with this email already exists');
    }

    // Create new user
    const hashedPassword = await hashPassword(dto.password);
    
    const user = this.userRepository.create({
      email: dto.email,
      password: hashedPassword,
      first_name: dto.first_name,
      last_name: dto.last_name,
      middle_name: dto.middle_name || null,
      role: UserRole.USER,
      is_verified: false, // Временно false, т.к. подтверждение почты не делаем
    });

    await this.userRepository.save(user);

    // TODO: Отправить email с подтверждением (заглушка)
    console.log(`[MOCK] Verification email sent to ${user.email}`);

    return this.toUserResponse(user);
  }

  async login(dto: LoginDto): Promise<{ user: UserResponse; tokens: TokenPairResponse }> {
    // Find user
    const user = await this.userRepository.findOne({
      where: { email: dto.email },
    });

    if (!user) {
      throw new Error('Invalid email or password');
    }

    // Check password
    const isPasswordValid = await checkPassword(dto.password, user.password);
    if (!isPasswordValid) {
      throw new Error('Invalid email or password');
    }

    // Generate tokens
    const tokens = this.generateTokens(user);

    return {
      user: this.toUserResponse(user),
      tokens,
    };
  }

  async refreshToken(refreshToken: string): Promise<TokenPairResponse> {
    const { verifyRefreshToken } = require('../../utils/jwt');
    const payload = verifyRefreshToken(refreshToken);

    if (!payload) {
      throw new Error('Invalid or expired refresh token');
    }

    // Find user to ensure they still exist
    const user = await this.userRepository.findOne({
      where: { id: payload.userId },
    });

    if (!user) {
      throw new Error('User not found');
    }

    // Generate new tokens
    return this.generateTokens(user);
  }

  async logout(refreshToken: string): Promise<void> {
    // В простой реализации JWT мы не храним токены,
    // поэтому logout просто говорит клиенту удалить токены.
    // Для полноценной инвалидации нужен черный список токенов в Redis.
    console.log(`[MOCK] Logout for token: ${refreshToken.substring(0, 20)}...`);
  }

  async getCurrentUser(userId: number): Promise<UserResponse> {
    const user = await this.userRepository.findOne({
      where: { id: userId },
    });

    if (!user) {
      throw new Error('User not found');
    }

    return this.toUserResponse(user);
  }
}